File: qt/gen_qfocusframe.go
package qt

import (
	"unsafe"
)

type QFocusFrame struct {
	h          uintptr
	isSubclass bool
}

// NewQFocusFrame constructs a new QFocusFrame object.
func NewQFocusFrame(parent *QWidget) *QFocusFrame {
	ret := newQFocusFrame(QFocusFrame_new(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQFocusFrame2 constructs a new QFocusFrame object.
func NewQFocusFrame2() *QFocusFrame {
	ret := newQFocusFrame(QFocusFrame_new2())
	ret.isSubclass = true
	return ret
}

func (this *QFocusFrame) MetaObject() *QMetaObject {
	return newQMetaObject(QFocusFrame_MetaObject(this.h))
}

func (this *QFocusFrame) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QFocusFrame_Metacast(this.h, param1_Cstring))
}

func QFocusFrame_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QFocusFrame_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QFocusFrame) SetWidget(widget *QWidget) {
	QFocusFrame_SetWidget(this.h, widget.cPointer())
}

func (this *QFocusFrame) Widget() *QWidget {
	return newQWidget(QFocusFrame_Widget(this.h))
}

func QFocusFrame_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QFocusFrame_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QFocusFrame_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QFocusFrame_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QFocusFrame) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QFocusFrame_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QFocusFrame) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QFocusFrame_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QFocusFrame_MetaObject
func miqt_exec_callback_QFocusFrame_MetaObject(self QFocusFrame, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QFocusFrame{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QFocusFrame) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QFocusFrame_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QFocusFrame) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QFocusFrame_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QFocusFrame_Metacast
func miqt_exec_callback_QFocusFrame_Metacast(self QFocusFrame, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QFocusFrame{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/gen_qfocusframe.cpp
// +build ignore

#include <QEvent>
#include <QFocusFrame>
#include <QMetaObject>
#include <QObject>
#include <QPaintDevice>
#include <QPaintEvent>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QStyleOption>
#include <QWidget>
#include <qfocusframe.h>
#include "gen_qfocusframe.h"
class MiqtVirtualQFocusFrame : public virtual QFocusFrame {
public:
MiqtVirtualQFocusFrame(QWidget* parent): QFocusFrame(parent) {};
MiqtVirtualQFocusFrame(): QFocusFrame() {};

virtual ~MiqtVirtualQFocusFrame() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QFocusFrame::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QFocusFrame_MetaObject(const_cast<MiqtVirtualQFocusFrame*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QFocusFrame::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QFocusFrame::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QFocusFrame_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QFocusFrame::qt_metacast(param1);

	}
};
QFocusFrame* QFocusFrame_new(QWidget* parent) {
return new MiqtVirtualQFocusFrame(parent);
}
QFocusFrame* QFocusFrame_new2() {
return new MiqtVirtualQFocusFrame();
}
void QFocusFrame_virtbase(QFocusFrame* src
, QWidget** outptr_QWidget
) {
*outptr_QWidget = static_cast<QWidget*>(src);
}
QMetaObject* QFocusFrame_MetaObject(const QFocusFrame* self) {
	return (QMetaObject*) self->metaObject();
}
void* QFocusFrame_Metacast(QFocusFrame* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QFocusFrame_Tr(const char* s) {
	QString _ret = QFocusFrame::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QFocusFrame_SetWidget(QFocusFrame* self, QWidget* widget) {
	self->setWidget(widget);
}
QWidget* QFocusFrame_Widget(const QFocusFrame* self) {
	return self->widget();
}
struct miqt_string QFocusFrame_Tr2(const char* s, const char* c) {
	QString _ret = QFocusFrame::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QFocusFrame_Tr3(const char* s, const char* c, int n) {
	QString _ret = QFocusFrame::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QFocusFrame_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQFocusFrame*>( (QFocusFrame*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QFocusFrame_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQFocusFrame*)(self) )->virtualbase_MetaObject();
}
void QFocusFrame_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQFocusFrame*>( (QFocusFrame*)(self) )->handle__Metacast = slot;
}
void* QFocusFrame_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQFocusFrame*)(self) )->virtualbase_Metacast(param1);
}
void QFocusFrame_Delete(QFocusFrame* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQFocusFrame*>( self );
	} else {
		delete self;
	}
}
